library(dplyr)
library(Seurat)
library(patchwork)
library(reshape2)
library(RColorBrewer)
library(scales)
library(ggplot2)
library(ggpubr)
library(ggplotify)
library(pheatmap)
Sys.setenv(LANGUAGE = "en") 
options(stringsAsFactors = FALSE) 

#fig9a,figS16a
#refer to fig2d

#fig9b
#refer to fig4c

#fig9c
library(ClusterGVis)
library(org.Mm.eg.db)
markers.all <- Seurat::FindAllMarkers(data.combined,
                                           only.pos = TRUE,
                                           min.pct = 0.25,
                                           logfc.threshold = 0.25)
markers <-markers.all %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(n = 50, wt = avg_log2FC)
st.data <- prepareDataFromscRNA(object = data.combined,
                                diffData = markers,
                                keep.uniqGene = T,
                                showAverage = TRUE)
enrich <- enrichCluster(object = st.data,
                        OrgDb = org.Mm.eg.db,
                        type = "BP",
                        organism = "mmu",
                        pvalueCutoff = 0.05,
                        topn = 8,
                        seed = 12345)
markers <- markers.all %>%
  dplyr::group_by(cluster) %>%
  dplyr::top_n(n = 5, wt = avg_log2FC)
markGenes= unique(markers$gene)
pdf('fig9e.pdf',height = 10,width = 16,onefile = F)
visCluster(object = st.data,
           plot.type = "both",
           column_names_rot = 45,
           show_row_dend = F,
           markGenes = markGenes,
           markGenes.side = "left",
           sample.col=sample_color,
           ctAnno.col=sample_color,
           annoTerm.data = enrich,
           line.side = "left",
           cluster.order = c(1:3),
)
dev.off()


#figS16b
degs <- FindAllMarkers(object = data.combined, 
                       only.pos = F,
                       logfc.threshold = 0.25) 
library(scRNAtoolVis)
markerVocalno(markers = degs,
              topn = 5,
              log2FC=0,
              labelCol = celltype_col
              )
ggsave("fig9d.pdf",width = 12,height = 5.5)


#fig9d,figS16c,d
#infercnv analysis
library(infercnv)
library(magrittr)
library(dplyr)
library(tidyverse)
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
library(AnnoProbe)
data.combined_raw=data.combined
for (i in unique(data.combined_raw$group)) {
   cell <- rownames(subset(data.combined_raw@meta.data, group == i))
  data.combined_raw$name=colnames(data.combined_raw)
  sub_CT<- subset(data.combined_raw, name %in% cell)
  data.combined=sub_CT
  data.combined$fincell=Idents(data.combined)
   meta=data.combined@meta.data
  groupFiles=meta[,c("name","fincell")]
  groupFiles_con=groupFiles[groupFiles$fincell%in%c("T_cell","B_cell"),]
  set.seed(666666)
  groupFiles_con=groupFiles_con[sample(rownames(groupFiles_con),3000),]
  groupFiles_con$fincell<-"cCON"
  groupFiles_test=groupFiles[groupFiles$fincell%in%c("Epi_Tumor"),]
  groupFiles=rbind(groupFiles_con,groupFiles_test)
  write.table(groupFiles,file = paste0(i,'_groupFiles.txt'),sep = '\t',quote = F,col.names = F,row.names = F)
  name_t=rownames(groupFiles)
  data.combined_fib <- subset(data.combined, name %in% as.vector(name_t))
  dat=as.matrix(data.combined_fib@assays$RNA@counts) 
 geneInfor=annoGene(rownames(dat),"SYMBOL",'mouse')
  colnames(geneInfor)
  geneInfor=geneInfor[with(geneInfor, order(chr, start)),c(1,4:6)]
  geneInfor=geneInfor[!duplicated(geneInfor[,1]),]
  length(unique(geneInfor[,1]))
  head(geneInfor)
  dat=dat[rownames(dat) %in% geneInfor[,1],]
  dat=dat[match( geneInfor[,1], rownames(dat) ),] 
  dim(dat)
  expFile=paste0(i,'_expFile.txt')
  write.table(dat,file = expFile,sep = '\t',quote = F)
  geneFile=paste0(i,'_geneFile.txt')
  write.table(geneInfor,file = geneFile,sep = '\t',quote = F,col.names = F,row.names = F)
   expFile=paste0(i,'_expFile.txt')
  groupFiles=paste0(i,'_groupFiles.txt')
  geneFile=paste0(i,'_geneFile.txt')
  infercnv_obj = CreateInfercnvObject(raw_counts_matrix=expFile,
                                      annotations_file=groupFiles,
                                      delim="\t",
                                      gene_order_file= geneFile,
                                      ref_group_names=c("cCON")
  )  
  infercnv_obj = infercnv::run(infercnv_obj,
                               cutoff=0.1, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics
                               out_dir=paste0("./",i), 
                               cluster_by_groups=F, 
                               num_threads=36,
                               analysis_mode = "subclusters",
                               denoise=TRUE,
                               no_plot = T,
                               HMM=T)
  }
#The figS16c was generated by UPhyloplot2 software
#Annotation the tree
library(tidyverse)
chr_pq <- read.csv("./chr_pq.new_mouse.csv", row.names=NULL)
chr_pq$chr=paste("chr",chr_pq$chr,sep = "")
cnv_regions=read.table("./HMM_CNV_predictions.HMMi6.rand_trees.hmm_mode-subclusters.Pnorm_0.5.pred_cnv_regions.dat",header = T,sep = "\t",stringsAsFactors = F)
cnv_regions=cnv_regions%>%filter(state!=3)
cnv_regions$cell_group_name=str_replace(cnv_regions$cell_group_name,"all.*observations\\.","")
cnv_regions$cnv_type=""
for (i in 1:nrow(cnv_regions)) {
  tmp_chr_pq=chr_pq%>%filter(chr==cnv_regions[i,"chr"])
  a=1
  while (cnv_regions[i,"start"] > tmp_chr_pq[a,"cutoff"] & a < length(tmp_chr_pq$cutoff)) {
    a=a+1
  }

  b=1
  while (cnv_regions[i,"end"] > tmp_chr_pq[b,"cutoff"] & b < length(tmp_chr_pq$cutoff)) {
    b=b+1
  }
  
  if (a==b) {
    cnv_regions[i,"cnv_type"] = paste(tmp_chr_pq[1,"chr"], tmp_chr_pq[a,"arm"],sep = "")
  }else {
    
    tmp = paste(cnv_regions[i,"chr"],tmp_chr_pq[a:b,"arm"],sep = "")
    tmp2 = as.vector(rbind(tmp, ","))
    tmp2=tmp2[-length(tmp2)]
    tmp3=paste(tmp2,collapse = "")
    cnv_regions[i,"cnv_type"] = tmp3
    
  }
  
  if (cnv_regions[i,"state"] < 3) {
    cnv_regions[i,"cnv_type"]=paste(cnv_regions[i,"cnv_type"],"_loss",sep = "")
  }
  if (cnv_regions[i,"state"] > 3) {
    cnv_regions[i,"cnv_type"]=paste(cnv_regions[i,"cnv_type"],"_gain",sep = "")
  }
  
  if (str_detect(cnv_regions[i,"cnv_type"],",")) {
    lenth=length(strsplit(cnv_regions[i,"cnv_type"],",")[[1]])
    tmp1=strsplit(cnv_regions[i,"cnv_type"],",")[[1]][1:(lenth-1)]
    tmp2=strsplit(strsplit(cnv_regions[i,"cnv_type"],",")[[1]][lenth],"_")[[1]][1]
    tmp3=strsplit(strsplit(cnv_regions[i,"cnv_type"],",")[[1]][lenth],"_")[[1]][2]
    tmp4=c(tmp1,tmp2)
    tmp5=paste(tmp4,"_",tmp3,sep = "")
    tmp6 = as.vector(rbind(tmp5, ","))
    tmp6=tmp6[-length(tmp6)]
    tmp7=paste(tmp6,collapse = "")
    cnv_regions[i,"cnv_type"] = tmp7
    
    rm(list = c("tmp1","tmp2","tmp3","tmp4","tmp5","tmp6","tmp7","tmp"))
  }
}

write.csv(cnv_regions,"cnv_regions.csv")
cnv_regions_copy=cnv_regions
cnv_regions=cnv_regions_copy
cnv_regions_part1=cnv_regions[str_detect(cnv_regions$cnv_type,","),]
cnv_regions_part2=cnv_regions[!str_detect(cnv_regions$cnv_type,","),]
cnv_regions_part1_new=as.data.frame(matrix(NA,ncol = ncol(cnv_regions_part1), nrow = nrow(cnv_regions_part1)*2 ))
colnames(cnv_regions_part1_new)=colnames(cnv_regions_part1)
for (i in 1:dim(cnv_regions_part1)[1]) {
  cnv_regions_part1_new[2*i-1,]=cnv_regions_part1[i,]
  cnv_regions_part1_new[2*i,]  =cnv_regions_part1[i,]
  
  tmp_chr_pq=chr_pq[chr_pq$chr == cnv_regions_part1[i,"chr"],]
  cnv_regions_part1_new[2*i-1,"end"] = tmp_chr_pq[1,"cutoff"]
  cnv_regions_part1_new[2*i,"start"] = tmp_chr_pq[2,"cutoff"]
  
  cnv_regions_part1_new[2*i-1,"cnv_type"] = strsplit(cnv_regions_part1_new[2*i-1,"cnv_type"],",")[[1]][1]
  cnv_regions_part1_new[2*i,"cnv_type"]   = strsplit(cnv_regions_part1_new[2*i,  "cnv_type"],",")[[1]][2]
}
cnv_regions=cnv_regions_part2 %>% rbind(cnv_regions_part1_new)
cnv_regions$chr=factor(cnv_regions$chr,levels = paste("chr",1:22,sep = ""))
cnv_regions=cnv_regions%>%arrange(cell_group_name,chr,start)
cnv_regions$region_len=cnv_regions$end-cnv_regions$start+1
cnv_regions=cnv_regions[,c("cell_group_name","cnv_type","region_len")]
cnv_regions=as.data.frame(cnv_regions%>%group_by(cell_group_name,cnv_type)%>%dplyr::summarize(all_region_len=sum(region_len)))
cnv_regions$final_label=""
ratio=0.4
for (i in 1:dim(cnv_regions)[1]) {
  arm=strsplit(cnv_regions[i,"cnv_type"],"_")[[1]][1]
  ref_len=chr_pq[paste(chr_pq$chr,chr_pq$arm,sep = "") == arm,"arm_len"]
  
  if(cnv_regions[i,"all_region_len"] >= ref_len*ratio){
    cnv_regions[i,"final_label"]=cnv_regions[i,"cnv_type"]
  }else{
    cnv_regions[i,"final_label"]=""
  }
}
cnv_regions$cnv_type=NULL
colnames(cnv_regions)[3]="cnv_type"
cnv_regions=cnv_regions%>%filter(cnv_type != "")
write.table(cnv_regions[,c("cell_group_name","cnv_type")],file = "CNVgroup_and_CNVtype_in_sampleA.txt",
            quote = F,sep = "\t",row.names = F,col.names = T)
cnv_1.1.1.1=as.data.frame(cnv_regions%>%filter(cell_group_name=="1.1.1.1"))[,"cnv_type"]
cnv_1.1.1.2=as.data.frame(cnv_regions%>%filter(cell_group_name=="1.1.1.2"))[,"cnv_type"]
cnv_1.1.2.1 =as.data.frame(cnv_regions%>%filter(cell_group_name=="1.1.2.1"  ))[,"cnv_type"]
cnv_1.1.2.2 =as.data.frame(cnv_regions%>%filter(cell_group_name=="1.1.2.2"  ))[,"cnv_type"]
cnv_1.2.1.1=as.data.frame(cnv_regions%>%filter(cell_group_name=="1.2.1.1"))[,"cnv_type"]
cnv_1.2.1.2=as.data.frame(cnv_regions%>%filter(cell_group_name=="1.2.1.2"))[,"cnv_type"]
cnv_1.2.2.1=as.data.frame(cnv_regions%>%filter(cell_group_name=="1.2.2.1"))[,"cnv_type"]
cnv_1.2.2.2=as.data.frame(cnv_regions%>%filter(cell_group_name=="1.2.2.2"))[,"cnv_type"]
cnv_1.1.1=intersect(cnv_1.1.1.1,cnv_1.1.1.2)
cnv_1.1.1.1_uniq=setdiff(cnv_1.1.1.1,cnv_1.1.1)
cnv_1.1.1.2_uniq=setdiff(cnv_1.1.1.2,cnv_1.1.1)
cnv_1.1.2=intersect(cnv_1.1.2.1,cnv_1.1.2.2)
cnv_1.1.2.1_uniq=setdiff(cnv_1.1.2.1,cnv_1.1.1)
cnv_1.1.2.2_uniq=setdiff(cnv_1.1.2.2,cnv_1.1.1)
cnv_1.2.1=intersect(cnv_1.2.1.1,cnv_1.2.1.2)
cnv_1.2.1.1_uniq=setdiff(cnv_1.2.1.1,cnv_1.2.1)
cnv_1.2.1.2_uniq=setdiff(cnv_1.2.1.2,cnv_1.2.1)
cnv_1.2.2=intersect(cnv_1.2.2.1,cnv_1.2.2.2)
cnv_1.2.2.1_uniq=setdiff(cnv_1.2.2.1,cnv_1.2.2)
cnv_1.2.2.2_uniq=setdiff(cnv_1.2.2.2,cnv_1.2.2)
cnv_1.1=intersect(cnv_1.1.1,cnv_1.1.2)
cnv_1.1.1_uniq=setdiff(cnv_1.1.1,cnv_1.1)
cnv_1.1.2_uniq=setdiff(cnv_1.1.2,cnv_1.1)
cnv_1.2=intersect(cnv_1.2.1,cnv_1.2.2)
cnv_1.2.1_uniq=setdiff(cnv_1.2.1,cnv_1.2)
cnv_1.2.2_uniq=setdiff(cnv_1.2.2,cnv_1.2)
cnv_1=intersect(cnv_1.1,cnv_1.2)
cnv_1.1_uniq=setdiff(cnv_1.1,cnv_1)
cnv_1.2_uniq=setdiff(cnv_1.2,cnv_1)
cat("cnv_1:",cnv_1,"\n",file="out.txt",append = T)
cat("cnv_1.1_uniq:",cnv_1.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2_uniq:",cnv_1.2_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.1.1_uniq:",cnv_1.1.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.1.2_uniq:",cnv_1.1.2_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2.1_uniq:",cnv_1.2.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2.2_uniq:",cnv_1.2.2_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.1.2.1_uniq:",cnv_1.1.2.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.1.2.2_uniq:",cnv_1.1.2.2_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.1.1.1_uniq:",cnv_1.1.1.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.1.1.2_uniq:",cnv_1.1.1.2_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2.1.1_uniq:",cnv_1.2.1.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2.1.2_uniq:",cnv_1.2.1.2_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2.2.1_uniq:",cnv_1.2.2.1_uniq,"\n",file="out.txt",append = T)
cat("cnv_1.2.2.2_uniq:",cnv_1.2.2.2_uniq,"\n",file="out.txt",append = T)

#fig9d
library(infercnv)
library(tidyverse)
library(ComplexHeatmap)
library(circlize)
library("RColorBrewer")
infercnv_obj = readRDS("./run.final.infercnv_obj")
expr <- infercnv_obj@expr.data
normal_loc <- infercnv_obj@reference_grouped_cell_indices
normal_loc <- normal_loc$cCON
test_loc <- infercnv_obj@observation_grouped_cell_indices
test_loc <- test_loc$test
anno.df=data.frame(
  CB=c(colnames(expr)[normal_loc],colnames(expr)[test_loc]),
  class=c(rep("normal",length(normal_loc)),rep("test",length(test_loc)))
)
head(anno.df)
gn <- rownames(expr)
geneFile <- read.table("./geneFile.txt", header=FALSE, row.names=NULL)
rownames(geneFile)=geneFile$V1
sub_geneFile <-  geneFile[intersect(gn,geneFile$V1),]
expr=expr[intersect(gn,geneFile$V1),]
head(sub_geneFile,4)
expr[1:4,1:4]
geneFile_sub=geneFile[rownames(expr),]
expr1=expr[,colnames(expr)%in%rownames(kmeans_df_s[kmeans_df_s$class=="e_NeOT",])]
expr2=expr[,colnames(expr)%in%rownames(kmeans_df_s[kmeans_df_s$class=="d_T",])]
expr3=expr[,colnames(expr)%in%rownames(kmeans_df_s[kmeans_df_s$class=="c_OCDC",])]
expr4=expr[,colnames(expr)%in%rownames(kmeans_df_s[kmeans_df_s$class=="b_DC",])]
expr5=expr[,colnames(expr)%in%rownames(kmeans_df_s[kmeans_df_s$class=="a_NS",])]
expr_lis=list(expr1,expr2,expr3,expr4,expr5)
plot_lis=list()
for (i in c(1:length(expr_lis))) {
  dat1=geneFile_sub[,2:4]
  dat1$cnv=rowMeans(expr_lis[[i]])
  dat1_up=dat1[dat1$cnv>1,]
  dat1_no=dat1[dat1$cnv==1,]
  dat1_down=dat1[dat1$cnv<1,]
  dat1_down$cnv=-dat1_down$cnv
  dat1_plot=list(dat1_up,dat1_no,dat1_down)
  plot_lis[[i]]=dat1_plot
}

extend_chromosomes = function(bed, chromosome, prefix = "zoom_") {
  zoom_bed = bed[bed[[1]] %in% chromosome, , drop = FALSE]
  zoom_bed[[1]] = paste0(prefix, zoom_bed[[1]])
  rbind(bed, zoom_bed)
}

cytoband = read.cytoband(species  = "mm10")
cytoband_df = cytoband$df
chromosome = cytoband$chromosome
xrange = c(cytoband$chr.len, cytoband$chr.len[c("chr1", "chr11")])
normal_chr_index = c(2:10,12:19)
zoomed_chr_index = c(1,11)
# normalize in normal chromsomes and zoomed chromosomes separately
sector.width = c(xrange[normal_chr_index] / sum(xrange[normal_chr_index]), 
                 xrange[zoomed_chr_index] / sum(xrange[zoomed_chr_index])) 
pdf(file = "fig9d.pdf",
    width = 7.5,height = 7.5)
#circos.par(start.degree = 90)
circos.initializeWithIdeogram(chromosome.index = paste0("chr",c(2:10,12:19,1,11)),
                              sector.width = sector.width,species = "mm10")
circos.track(ylim = c(0, 1),bg.col = 'white', 
             bg.border = 'white', 
             track.height = 0.05)
circos.genomicTrackPlotRegion(plot_lis[[1]], 
                              ylim = c(-1,1.5), 
                              track.height = 0.1,                           
                              panel.fun = function(region, value, ...) {
                                circos.genomicLines(region, value, type = "h",baseline = 0, border =  if (value>1) {"#F36F75"
                                }else if(value<1){"#73A2F5"}else{"#F3CAC8"},lwd = 0.1,area = F,col=if (value>1) {"#F36F75"
                                }else{"#73A2F5"})},
                              bg.col = '#F3CAC8', 
                              bg.border = '#F3CAC8'
)

circos.genomicTrackPlotRegion(plot_lis[[2]], 
                              ylim = c(-1,1.5),
                              track.height = 0.1,
                              panel.fun = function(region, value, ...) {
                                circos.genomicLines(region, value, type = "h",baseline = 0, border =  if (value>1) {"#F36F75"
                                }else if(value<1){"#73A2F5"}else{"#F9DBB6"},lwd = 0.1,area = F,col=if (value>1) {"#F36F75"
                                }else{"#73A2F5"})},
                              bg.col = '#F9DBB6', 
                              bg.border = '#F9DBB6'
)
circos.genomicTrackPlotRegion(plot_lis[[3]], 
                              ylim = c(-1,1.5),
                              track.height = 0.1, 
                              panel.fun = function(region, value, ...) {
                                circos.genomicLines(region, value, type = "h",baseline = 0, border =  if (value>1) {"#F36F75"
                                }else if(value<1){"#73A2F5"}else{"#B2E6D4"},lwd = 0.1,area = F,col=if (value>1) {"#F36F75"
                                }else{"#73A2F5"})},
                              bg.col = '#B2E6D4', 
                              bg.border = '#B2E6D4'
)

circos.genomicTrackPlotRegion(plot_lis[[4]], 
                              ylim = c(-1,1.5),
                              track.height = 0.1,
                             panel.fun = function(region, value, ...) {
                                circos.genomicLines(region, value, type = "h",baseline = 0, border =  if (value>1) {"#F36F75"
                                }else if(value<1){"#73A2F5"}else{"#CBD5E3"},lwd = 0.1,area = F,col=if (value>1) {"#F36F75"  
                                }else{"#73A2F5"})},
                              bg.col = '#CBD5E3', 
                              bg.border = '#CBD5E3'
)
circos.genomicTrackPlotRegion(plot_lis[[5]], 
                              ylim = c(-1,1.5),
                              track.height = 0.1,
                              panel.fun = function(region, value, ...) {
                                circos.genomicLines(region, value, type = "h",baseline = 0, border =  if (value>1) {"#F36F75"
                                }else if(value<1){"#73A2F5"}else{"#DCD0E3"},lwd = 0.1,area = F,col=if (value>1) {"#F36F75"
                                }else{"#73A2F5"})},
                              bg.col = '#DCD0E3', 
                              bg.border = '#DCD0E3'
)
circos.clear()
dev.off()

#figS16d
sam=c("a_ns","b_dc","c_ocdc",
       "d_t","e_neot")
alltype=c()
for (t in sam) {
  group_cnvtype_raw=read.table(paste0("./",t,"/",r,"/CNVgroup_and_CNVtype_in_sampleA.txt"),header = T,sep = "\t",stringsAsFactors = F)
  alltype=c(alltype,group_cnvtype_raw$cnv_type)
}
alltype=unique(alltype)
some.sample.stat=data.frame(rep("tmp",length(alltype)),1)
for (tt in sam) {
  dir=paste0("./plot/",tt,".cell_groupings")
  cell_group=read.csv(dir,header = T,sep = "\t",stringsAsFactors = F)
  cell_group=cell_group[!str_detect(cell_group$cell_group_name,"references"),]
  cell_group$cell_group_name=str_replace(cell_group$cell_group_name,"all.*observations\\.","")
  group_cellcount=as.data.frame(table(cell_group$cell_group_name))
  colnames(group_cellcount)=c("cell_group_name","cellcount")
  group_cellcount$cellratio=group_cellcount$cellcount / sum(group_cellcount$cellcount)
  group_cnvtype=read.table(paste0("./",tt,"/",r,"/CNVgroup_and_CNVtype_in_sampleA.txt"),header = T,sep = "\t",stringsAsFactors = F)
  group_cnvtype=group_cnvtype%>%inner_join(group_cellcount,by="cell_group_name")
  cellpercent=c()
  for (i in alltype) {
    if(i %in% unique(group_cnvtype$cnv_type)){
      tmp=sum(group_cnvtype[group_cnvtype$cnv_type == i,"cellratio"])
      cellpercent=append(cellpercent,tmp)
    }else{
      cellpercent=append(cellpercent,0)
    }
  }
  names(cellpercent)=alltype
  one.sample.stat=as.data.frame(cellpercent)
   some.sample.stat[[tt]]=one.sample.stat
  }
some.sample.stat=as.data.frame(some.sample.stat)
rownames(some.sample.stat)=alltype
colnames(some.sample.stat)=sam
library(RColorBrewer)
library(scales)
library(pheatmap)
pdf(paste0("heat_cnv.pdf"),4,8)
pheatmap(some.sample.stat,
         cluster_rows = T,cluster_cols = F,
         color = colorRampPalette(brewer.pal(9,"PuRd"))(100),
         show_rownames=T,
         border_color = "grey"
         
)
dev.off()

#fig9e
library(CellChat)
library(patchwork)
library(ktplots)
library(SingleCellExperiment)
library(reshape2)
library(Seurat)
library(circlize)
library(igraph)
library(ggraph)
library(ggrepel)
library(RColorBrewer)
library(grid)
library(ggplot2)
for (i in c("a_NS","b_other","e_NeOT")) {
  data.input = data.combined@assays$RNA@data # normalized data matrix
  meta = data.combined@meta.data # a dataframe with rownames containing cell mata data
  cell.use = rownames(meta)[meta$group == i] # extract the cell names from disease data
  # Prepare input data for CelChat analysis
  data.input = data.input[, cell.use]
  meta = meta[cell.use, ]
  cellchat <- createCellChat(object = data.input, meta = meta, group.by = "celltype")
  cellchat <- addMeta(cellchat, meta = meta)
  cellchat <- setIdent(cellchat, ident.use = "fincell") # set "labels" as default cell identity
   groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group
  CellChatDB <- CellChatDB.mouse # use CellChatDB.mouse if running on mouse data
  CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling
  cellchat@DB <- CellChatDB
  # subset the expression data of signaling genes for saving computation cost
  cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
  future::plan("multiprocess", workers = 12) # do parallel
  cellchat <- identifyOverExpressedGenes(cellchat)
  cellchat <- identifyOverExpressedInteractions(cellchat)
  # project gene expression data onto PPI network (optional)
  cellchat <- projectData(cellchat, PPI.mouse)
  cellchat <- computeCommunProb(cellchat)
  # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
  cellchat <- filterCommunication(cellchat, min.cells = 10)
  cellchat <- computeCommunProbPathway(cellchat)
    cellchat <- aggregateNet(cellchat)
  saveRDS(cellchat, file = paste0("cellchat_",i,".rds"))
  
}
i="Malignant"
 gg1 <- netVisual_bubble(cellchat, sources.use = i, targets.use = c(1:20),  comparison = c(1, 2), max.dataset = 2, title.name = "Increased signaling", angle.x = 45, remove.isolate = T)
dat=gg1$data
path=paste0(dat$source,dat$target,dat$ligand,dat$receptor)
dat$path=path
write.csv(dat,"malignant_source_up.csv")
plot.data$from <- paste0(plot.data$ligand, "_", plot.data$source)
plot.data$to <- paste0(plot.data$receptor, "_", plot.data$target)
plot.data$logFC <- limit(plot.data$logFC, min = -2, max = 3)
plot.data$logpadj <- limit(plot.data$logpadj, min = 10, max = 20)
plot.data <- subset(plot.data, L_Fraction>min.frac, ) 
if (!is.null(source.group)) plot.data <- subset(plot.data, source %in% source.group)
if (!is.null(target.group)) plot.data <- subset(plot.data, target %in% target.group)
vertices <- data.frame(
  name = c(plot.data$from, plot.data$to),
  celltype = c(plot.data$source, plot.data$target),
  type = rep(c("ligand", "receptor"), each = nrow(plot.data)),
  frac = c(plot.data$L_Fraction, plot.data$R_Fraction),
  label = c(plot.data$ligand, plot.data$receptor),
  logpadj = c(plot.data$logpadj, plot.data$logpadj),
  logFC = c(plot.data$logFC, plot.data$logFC)
)
vertices <- vertices[!duplicated(vertices$name), ]
vertices$cellident <- paste0(vertices$celltype, "_", vertices$type)
vertices.head <- data.frame(matrix(ncol = ncol(vertices), 
                                   nrow = length(unique(c(vertices$celltype, vertices$cellident)))+1))
colnames(vertices.head) <- colnames(vertices)
vertices.head$name <- c("root", unique(vertices$celltype), unique(vertices$cellident))
vertices <- rbind(vertices.head[, colnames(vertices)], vertices)

edgelist <- rbind(
  data.frame("from" = paste0(plot.data$source, "_ligand"), "to" = plot.data$from),
  data.frame("from" = paste0(plot.data$target, "_receptor"), "to" = plot.data$to)
)
edgelist.head <- data.frame("to" = unique(na.omit(vertices$cellident)))
edgelist.head$from <- gsub("_receptor|_ligand", "", edgelist.head$to)
edgelist.head <- rbind(edgelist.head, data.frame("from" = "root", "to" = unique(na.omit(vertices$celltype))))
edgelist.head[, c("logFC", "logpadj")] = NA
edgelist <- rbind(edgelist.head[, colnames(edgelist)], edgelist)
mygraph <- graph_from_data_frame(d = edgelist, directed = T, vertices=vertices)
d <- igraph::as_data_frame(mygraph, "both")
from  <-  match(plot.data$from, vertices$name)
to  <-  match(plot.data$to, vertices$name)
e.fun <- get_con(from = from, to = to)
e.df <- e.fun(create_layout(mygraph, layout = "dendrogram"))
e.df$logpadj[is.na(e.df$logpadj)] <- 0
if (is.null(cell.col)){
  cell = intersect(unique(seu@meta.data[[celltypes]]), vertices$celltype)
  cell.col <- setNames(object = brewer.pal(length(cell), "Set3"),
                       nm = cell)
}
ggraph(mygraph, layout = 'dendrogram', circular = TRUE) +
  geom_conn_bundle(data = get_con(from = from, to = to),
                   aes(colour = logFC), width = 1.8, tension = 1,
                   arrow = arrow(length = unit((e.df$logpadj)*0.2, "mm"))) +
  geom_node_point(aes(size = frac, filter = leaf,color = celltype,shape = type)) + 
  geom_text_repel(aes(x = x, y = y, label = label), 
                  segment.square = TRUE, segment.inflect = TRUE, 
                  segment.size = 0.2, force = 0.5, size = 3, 
                  force_pull = 0) +
  scale_edge_colour_gradientn(colors = c("white", "pink", "#B91129"), limits = c(0,2)) + 
  scale_shape_manual(values = c("ligand" = 1, "receptor" = 19)) +   
  scale_color_manual(values = col) +                           
  theme_void()
ggsave("fig9e.pdf",width = 6,height = 5.5)

#fig9g
library(R.utils)
library(GenomicFeatures)
library(ChIPseeker)
library(ggplot2)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(Seurat)
library(dplyr)
library(patchwork)
library(Matrix)
library(data.table)
txdb=TxDb.Mmusculus.UCSC.mm10.knownGene
test_sam=c("T","DC","NeOT","NS","OCDC")
neo=neo_posistion

for (sam in test_sam) {

 tmp= read.csv(paste0("pass2_position_Malignant_",sam,".csv"), row.names=1)
  pos=data_frame(tmp[,1],as.numeric(tmp[,2]),as.numeric(tmp[,3]))
 
 colnames(pos)=c("chr","start","end")
 pos$end=pos$start+1
 pos=as.data.frame(pos)
 peak=GRanges(seqnames = Rle(pos[,1]),
              ranges = IRanges(pos[,2],pos[,3]),strand = rep(c("*"),nrow(pos)))
 
  peakAnnoList <- annotatePeak(peak, tssRegion=c(-3000,3000), TxDb=txdb,annoDb = "org.Mm.eg.db")
 pos_ann=as.data.frame(peakAnnoList)
 tmp=pos_ann$annotation
 tmp=strsplit(tmp," ")
 tmp=lapply(tmp,FUN = function(x){x[1]})
 tmp=unlist(tmp)
 names(tmp)=NULL
 pos_ann$Type=tmp
 pos_ann$Shape="circle"
 pos_ann[pos_ann$Type=="Promoter",]$Shape="triangle"
 pos_ann[pos_ann$Type=="Exon",]$Shape="circle"
 pos_ann[pos_ann$Type=="Intron",]$Shape="circle"
 pos_ann[pos_ann$Type=="3'",]$Shape="box"
 pos_ann[pos_ann$Type=="5'",]$Shape="box"
 pos_ann[pos_ann$Type=="Distal",]$Shape="box"
 pos_ann[pos_ann$Type=="Downstream",]$Shape="box"
 pos_ann$Chr=gsub("chr","",pos_ann$seqnames)
 pos_ann$Start=pos_ann$start
 pos_ann$End=pos_ann$end
 pos_ann$color="#33A02A"
 pos_ann[pos_ann$Type=="Promoter",]$color="#D1AA63"
 pos_ann[pos_ann$Type=="Exon",]$color="#FF7F00"
 pos_ann[pos_ann$Type=="Intron",]$color="#33A02A"
 pos_ann[pos_ann$Type=="3'",]$color="#1CC5FE"
 pos_ann[pos_ann$Type=="5'",]$color="#60AFB3"
 pos_ann[pos_ann$Type=="Distal",]$color="#BBD4E0"
 pos_ann[pos_ann$Type=="Downstream",]$color="#989898"
 pos_ann$index=paste0(pos_ann$seqnames,"_",pos_ann$start)
 pos_ann[pos_ann$index%in%neo,]$color="#962120"
 write.csv(pos_ann,paste0("ann2_",sam,".csv"))
}
driver_gene=driver_luad
gene_neo=c("PANK3","TMEM101","RAB3GAP2","NEURL4","SUPT6")
neo_pos=c("chr11_35783472","chr11_102155755","chr1_185263697","chr11_69908838","chr11_78222422")

for (sam in test_sam) {
 tmp=read.csv(paste0("ann2_",sam,".csv"), row.names=1)
 
 tmp$SYMBOL2=toupper(tmp$SYMBOL)
 tmp=tmp[tmp$SYMBOL2%in%c(driver_gene$SYMBOL,gene_neo),]
 row.names(driver_gene)=driver_gene$SYMBOL
 type=driver_gene[tmp$SYMBOL2,]$ROLE
 type[is.na(type)]="neo_gene"
 tmp$Type=type
 tmp[tmp$index%in%neo_pos,]$Type="neo_pos"
 tmp$color="color"
 tmp[tmp$Type=="Act",]$color="FF7F00"
 tmp[tmp$Type=="LoF",]$color="33A02A"
 tmp[tmp$Type=="ambiguous",]$color="CCCCCC"
 tmp[tmp$Type=="neo_gene",]$color="FB7D80"
 tmp[tmp$Type=="neo_pos",]$color="462672"
 write.csv(tmp,paste0("plot2_",sam,".csv"))
}

library(RIdeogram)

gene_density <- GFFex(input = "gencode.vM25.annotation.gff3.gz", 
                      karyotype = "mmu_karyotype.txt", 
                      feature = "gene", window = 1500000)

mmu_karyotype <- read.table("mmu_karyotype.txt", 
                            sep = "\t", header = T, 
                            stringsAsFactors = F)
test_sam=c("T","DC","NeOT","NS","OCDC")
sam=test_sam[1]
tmp= read.csv(paste0("plot2_",sam,".csv"), row.names=1)
row.names(tmp)=tmp$index
label=tmp[,18:23]
label=label[label$Chr%in%paste0(c(1:19,"X","Y")),]
label$Chr=paste0("chr",label$Chr)
#label$chr=label$Chr
#label=label[,-7]
colnames(label)[3]="Chr"
for (i in neo_pos) {
  if (is.na(label[i,1])) {
    label[i,]=c("neo_pos","triangle",unlist(strsplit(i,"_"))[1],as.numeric(unlist(strsplit(i,"_"))[2]),as.numeric(as.numeric(unlist(strsplit(i,"_"))[2])+1),"462672")
  }
}
label=label[order(label$Chr),]
ideogram(karyotype = mmu_karyotype, overlaid = gene_density,
         label = xx,label_type="marker",Lx = 180,
         colorset1 = c("white","#6699CC","#FA9A99","#A60000"))
convertSVG("fig9g.svg", device = "png")

#fig9g
library(ggplot2)
library(dplyr)
library(ggalluvial)
mycolor=c("#FF7F00",
            "#1F78B4", 
            "#33A02C",
            "#6A3D9A", 
            "#E31A1C"
)
plot$group=factor(plot$group,levels = c("NS","DC","OCDC","T","NeOT"))
ggplot(plot,aes(x=group,y=ratio,fill=gene,stratum=gene,alluvium=gene))+
  geom_col(width = 0.6,color="black")+
  geom_flow(width = 0.5,alpha=0.4,knot.pos = 0.1)+
  scale_fill_manual(values = mycolor)+
  theme_classic()+ coord_flip()
































